driver-async/src/main/com/mongodb/async/client/gridfs/GridFSDownloadStreamImpl.java

```diff
@@ -122,61 +122,87 @@ public void onResult(final GridFSFile result, final Throwable t) {
-            amountToRead = (int) fileInfo.getLength();
-        }
-        read(amountToRead, dst, errorHandlingCallback);
+        checkAndFetchResults(0, dst, new SingleResultCallback<Integer>() {
+            @Override
+            public void onResult(final Integer result, final Throwable t) {
+                releaseReadingLock();
+                errorHandlingCallback.onResult(result, t);
+            }
+        });
     }
 
-    private void read(final int amountRead, final ByteBuffer dst, final SingleResultCallback<Integer> callback) {
+    private void checkAndFetchResults(final int amountRead, final ByteBuffer dst, final SingleResultCallback<Integer> callback) {
         if (currentPosition == fileInfo.getLength() || dst.remaining() == 0) {
-            releaseReadingLock();
             callback.onResult(amountRead, null);
-            return;
-        }
-
-        boolean fetchBuffer = false;
-        if (buffer == null) {
-            fetchBuffer = true;
-        } else if (bufferOffset == buffer.length) {
-            chunkIndex += 1;
-            bufferOffset = 0;
-            fetchBuffer = true;
-        }
-
-        final RecursiveReadCallback recursiveCallback = new RecursiveReadCallback(amountRead, dst, callback);
-        if (fetchBuffer) {
-            getBuffer(chunkIndex, new SingleResultCallback<byte[]>() {
+        } else if (!resultsQueue.isEmpty()) {
+            processResults(amountRead, dst, callback);
+        } else if (cursor == null) {
+            chunksCollection.find(new Document("files_id", fileInfo.getId())
+                    .append("n", new Document("$gte", chunkIndex)))
+                    .batchSize(batchSize).sort(new Document("n", 1))
+                    .batchCursor(new SingleResultCallback<AsyncBatchCursor<Document>>() {
+                        @Override
+                        public void onResult(final AsyncBatchCursor<Document> result, final Throwable t) {
+                            if (t != null) {
+                                callback.onResult(null, t);
+                            } else if (result == null) {
+                                chunkNotFound(chunkIndex, callback);
+                            } else {
+                                cursor = result;
+                                checkAndFetchResults(amountRead, dst, callback);
+                            }
+                        }
+                    });
+        } else {
+            cursor.next(new SingleResultCallback<List<Document>>() {
                 @Override
-                public void onResult(final byte[] result, final Throwable t) {
+                public void onResult(final List<Document> result, final Throwable t) {
                     if (t != null) {
-                        releaseReadingLock();
                         callback.onResult(null, t);
+                    } else if (result == null || result.isEmpty()) {
+                        chunkNotFound(chunkIndex, callback);
```

Comment:
This is an unexpected state, since we're relying on the fileLength check to stop fetching more data rather than end of cursor.  The only way to test this logic would be to manually remove the last chunk.